editoast/src/views/timetable/train_schedule.rs

```diff
@@ -590,10 +608,348 @@ async fn occupancy_blocks(
+                    time_begin: train_schedule.start_time + Duration::milliseconds(time as i64),
+                    duration: PositiveDuration::default(),
+                    train_schedule_id: train_schedule.id,
+                },
+            ));
+        }
+    }
+    None
+}
+
+fn find_track_occupancy_for_operational_point(
+    operational_point_id: &str,
+    operational_point_track_offsets: &[TrackOffset],
+    path_item_cache: &PathItemCache,
+    simulation: &Response,
+    pathfinding: &PathfindingResult,
+    train_schedule: &models::TrainSchedule,
+) -> Option<(String, TrackOccupancy)> {
+    // Get the track ranges from the pathfinding
+    let (track_ranges, path_item_positions) = match pathfinding {
+        PathfindingResult::Success(pathfinding_result_success) => {
+            assert_eq!(
+                pathfinding_result_success.path_item_positions.len(),
+                train_schedule.path.len()
+            );
+            (
+                &pathfinding_result_success.track_section_ranges,
+                &pathfinding_result_success.path_item_positions,
+            )
+        }
+        PathfindingResult::Failure(_) => {
+            // The timetable can have a pathfinding failure for a train schedule
+            // because the train schedule is not valid.
+            // We skip it because we don't want to compute the track occupancy for an invalid train schedule.
+            // It's not a problem because we are just looking for the track occupancy of the operational point for the valid trains.
+            tracing::info!(train_schedule.id, "pathfinding failed");
+            return None;
+        }
+    };
+    let path_projection = PathProjection::new(track_ranges);
+
+    // Get the positions and the times from the simulation
+    let report_train = match simulation {
+        Response::Success { final_output, .. } => final_output.report_train.clone(),
+        _ => {
+            tracing::info!(train_schedule.id, "simulation failed");
+            return None;
+        }
+    };
+
+    let path_item_id =
+        match_path_item_id_with_op_id(path_item_cache, train_schedule, operational_point_id);
+    if let Some(path_item_id) = path_item_id {
+        track_occupancy_on_path_item(
+            train_schedule,
+            path_item_id,
+            path_item_positions,
+            &path_projection,
+            operational_point_track_offsets,
+            &report_train,
+        )
+    } else {
+        interpolate_track_occupancy(
+            train_schedule,
+            &path_projection,
+            operational_point_track_offsets,
+            &report_train,
+        )
+    }
+}
+
+/// Match the id of an operational point with a path item in the train schedule and return the id of the path item concerned
+fn match_path_item_id_with_op_id<'a>(
+    path_item_cache: &'a PathItemCache,
+    train_schedule: &'a models::TrainSchedule,
+    operational_point_id: &'a str,
+) -> Option<&'a str> {
```

Comment:
nitpick: the lifetime is too constrained from what I understand. You only need to tie the lifetime of the returned value to `train_schedule`.

```suggestion
fn match_path_item_id_with_op_id<'a>(
    path_item_cache: &PathItemCache,
    train_schedule: &'a models::TrainSchedule,
    operational_point_id: &str,
) -> Option<&'a str> {
```